src/artifacts_composefs.c

```diff
@@ -0,0 +1,318 @@
+#include <asm-generic/errno-base.h>
+#include <errno.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <glib/gstdio.h>
+#include <libcomposefs/lcfs-writer.h>
+
+#include "artifacts_composefs.h"
+#include "utils.h"
+
+static gint strcmp0_p(gconstpointer a, gconstpointer b)
+{
+	const gchar *str1 = *((gchar **) a);
+	const gchar *str2 = *((gchar **) b);
+
+	return g_strcmp0(str1, str2);
+}
+
+/*
+ * To access the objects in the image in sequential order, we need to sort them
+ * by name.
+ */
+static GPtrArray *get_objects_sorted(GHashTable *objects)
+{
+	GPtrArray *result = g_ptr_array_sized_new(g_hash_table_size(objects));
+	gchar *object_name = NULL;
+
+	GHashTableIter iter;
+	g_hash_table_iter_init(&iter, objects);
+	while (g_hash_table_iter_next(&iter, (gpointer*)&object_name, NULL)) {
+		g_ptr_array_add(result, object_name);
+	}
+
+	g_ptr_array_sort(result, strcmp0_p);
+
+	return result;
+}
+
+static void composefs_recurse_objects(GHashTable *objects, struct lcfs_node_s *node)
+{
+	g_return_if_fail(objects);
+	g_return_if_fail(node);
+
+	for (size_t i = 0; i < lcfs_node_get_n_children(node); i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		g_assert(child != NULL);
+
+		uint32_t type = lcfs_node_get_mode(child) & S_IFMT;
+		const gchar *payload = lcfs_node_get_payload(child);
+		if (type == S_IFREG && payload)
+			g_hash_table_add(objects, g_strdup(payload));
+
+		composefs_recurse_objects(objects, child);
+	}
+}
+
+static gboolean composefs_objects_from_image(GHashTable *objects, const gchar *image_path, GError **error)
+{
+	GError *ierror = NULL;
+	g_autofree gchar *image_bytes = NULL;
+	gsize image_length = 0;
+
+	g_return_val_if_fail(objects, FALSE);
+	g_return_val_if_fail(image_path, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* load image into memory */
+	if (!g_file_get_contents(image_path, &image_bytes, &image_length, &ierror)) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	struct lcfs_node_s *node = lcfs_load_node_from_image((uint8_t *)image_bytes, image_length);
+	composefs_recurse_objects(objects, node);
+	g_clear_pointer(&node, lcfs_node_unref);
+
+	return TRUE;
+}
+
+gboolean r_artifact_repo_prepare_composefs(RArtifactRepo *repo, GError **error)
```

Comment:
Function names do not really match the module name, but maybe at least use consistent prefixes?

`r_composefs_artifact_*` vs `r_artifact_repo_..._composefs`